Add description of basic routing logic of Kong Gateway

hanshuebner opened the pull request:
### Summary

Describe the basic routing strategy of Kong Gateway

### Reason

Some users don't understand how Kong Gateway routes requests and make incorrect assumptions.

https://github.com/Kong/kong/issues/9914#issuecomment-1349975871
https://github.com/Kong/kong/issues/9589

KAG-247

Guaris commented:
How do you feel about the info being contained in either of these docs and then linked the 
https://docs.konghq.com/gateway/latest/key-concepts/routes/
https://docs.konghq.com/gateway/latest/key-concepts/services/

app/_src/gateway/get-started/services-and-routes.md

hanshuebner commented:
> How do you feel about the info being contained in either of these docs https://docs.konghq.com/gateway/latest/key-concepts/routes/ https://docs.konghq.com/gateway/latest/key-concepts/services/ and then linked here: app/_src/gateway/get-started/services-and-routes.md

I moved the section into [key-concepts/services/](https://deploy-preview-4934--kongdocs.netlify.app/gateway/3.1.x/key-concepts/routes/#how-kong-gateway-routes-requests) and added a link to [get-started/services-and-routes/](https://deploy-preview-4934--kongdocs.netlify.app/gateway/3.1.x/key-concepts/routes/)

kikito commented on app/_src/gateway/key-concepts/routes/index.md:
```diff
@@ -15,6 +15,37 @@ For example, if you have an external application and an internal application tha
+before those with a lower `regex_priority`.  If routes have the same
+`regex_priority`, regular expressions with a lower (older)
+`created_at` timestamp in the database are considered before those
+that have a later `created_at` timestamp.
```
Perhaps we should note the exception on dbless mode.
```suggestion
that have a later `created_at` timestamp.  Note that on DB-less mode all entities have the same value for `created_at`, so in DB-less mode only `regex_priority` is used for sorting - routes with the same `regex_priority` 
```

cloudjumpercat commented on app/_src/gateway/key-concepts/routes/index.md:
```diff
@@ -15,6 +15,96 @@ For example, if you have an external application and an internal application tha
+Prior versions used the
+[PCRE library](https://www.pcre.org/original/doc/html/pcrepattern.html)
+to evaluate regular expression.  While the two engines are largely
+compatible, subtle differences exist between the two.  Please refer to
```
```suggestion
compatible, subtle differences exist between the two. Refer to
```

cloudjumpercat commented on app/_src/gateway/key-concepts/routes/index.md:
```diff
@@ -15,6 +15,96 @@ For example, if you have an external application and an internal application tha
+throughput and latency of {{site.base_gateway}}.
+
+Starting with version 3.0, {{site.base_gateway}} uses the regular
+expression engine shipped with the Rust programming language if the
```
```suggestion
expression engine shipped with the [Rust](https://docs.rs/regex/latest/regex/) programming language if the
```

cloudjumpercat reviewed (changes requested):
Left a few suggestions, thanks for submitting these changes!

kikito commented on app/_src/gateway/key-concepts/routes/index.md:
```diff
@@ -15,6 +15,96 @@ For example, if you have an external application and an internal application tha
+
+For each incoming request, {{site.base_gateway}} needs to determine
+which service gets to handle it based on the routes that have been
+defined.  We describe how the routing process works in
+`traditional_compat` mode, in a logical sense, below.  The actual
+router implementation optimizes the process and caches results to
+avoid having to scan all routes repeatedly.  Note that when using the
+`expressions` router, route priorities are explicitly defined and this
+description does not apply.  Underneath, routes are converted from the
+`traditional_compat` format to the `expressions` format.
```
I believe the first phrase of this amendment is incomplete (main verb is missing):

> The `traditional_compat` mode routing process *is described* in the following sections.

kikito commented on app/_src/gateway/key-concepts/routes/index.md:
```diff
@@ -15,6 +15,96 @@ For example, if you have an external application and an internal application tha
+priority, it is not defined which of the two routes will be used.
+
+In `traditional_compat` mode, the priority of a route is determined as
+follows, by the order of descending significance:
```
Perhaps use an ordered list instead of an unordered one here?

```
1. Priority points
2. Wildcard hosts
3. Header count
4. Regular expressions and prefix paths
```

kikito commented on app/_src/gateway/key-concepts/routes/index.md:
```diff
@@ -15,6 +15,96 @@ For example, if you have an external application and an internal application tha
+In `traditional_compat` mode, the priority of a route is determined as
+follows, by the order of descending significance:
+
+### Priority Points
```
```suggestion
### 1. Priority points
```

kikito commented on app/_src/gateway/key-concepts/routes/index.md:
```diff
@@ -15,6 +15,96 @@ For example, if you have an external application and an internal application tha
+`headers` defined, the second one will be considered before the first
+one.
+
+### Wildcard Hosts
```
```suggestion
### 2. Wildcard hosts
```

kikito commented on app/_src/gateway/key-concepts/routes/index.md:
```diff
@@ -15,6 +15,96 @@ For example, if you have an external application and an internal application tha
+any wildcard host specification will be considered after routes that
+do not have any wildcard host (or no host) specified.
+
+### Header Count
```
```suggestion
### 3. Header count
```

kikito commented on app/_src/gateway/key-concepts/routes/index.md:
```diff
@@ -15,6 +15,96 @@ For example, if you have an external application and an internal application tha
+specified headers have higher priority than those with a lower number
+of headers.
+
+### Regular Expressions and Prefix Paths
```
```suggestion
### 4. Regular Expressions and Prefix Paths
```

kikito reviewed (commented):
I like this doc. I would explicitly add numbers (ordered lists, header titles) to make it even more obvious what the order of evaluation is. This is a comment but not at all a blocker.

hanshuebner commented on app/_src/gateway/key-concepts/routes/index.md:
```diff
@@ -15,6 +15,96 @@ For example, if you have an external application and an internal application tha
+
+In general, the router uses the highest priority matching route to
+handle a request.  If there are multiple matching routes with the same
+priority, it is not defined which of the two routes will be used.
```
"Undefined" means that either of the routes will be chosen at the discretion of the implementation and we don't document which of the two it will be.  We used to use the "last change" timestamp of the route to enforce a particular ordering in the past, but as all those timestamps will be identical in hybrid mode or dbless, they are not useful as an additional sort criterion anymore.

cloudjumpercat commented on app/_src/gateway/key-concepts/routes/index.md:
```diff
@@ -15,6 +15,96 @@ For example, if you have an external application and an internal application tha
+3. Header count
+4. Regular expressions and prefix paths
+
+### 1. Priority points
```
```suggestion
### Priority points
```

cloudjumpercat commented on app/_src/gateway/key-concepts/routes/index.md:
```diff
@@ -15,6 +15,96 @@ For example, if you have an external application and an internal application tha
+`headers` defined, the second one will be considered before the first
+one.
+
+### 2. Wildcard hosts
```
```suggestion
### Wildcard hosts
```

cloudjumpercat commented on app/_src/gateway/key-concepts/routes/index.md:
```diff
@@ -15,6 +15,96 @@ For example, if you have an external application and an internal application tha
+any wildcard host specification will be considered after routes that
+don't have any wildcard host (or no host) specified.
+
+### 3. Header count
```
```suggestion
### Header count
```

hanshuebner commented on app/_src/gateway/key-concepts/routes/index.md:
```diff
@@ -15,6 +15,96 @@ For example, if you have an external application and an internal application tha
+`expressions` router, route priorities are explicitly defined and this
+description doesn't apply. When you use the 'expressions' router,
+routes are converted from the `traditional_compat` format to the
+`expressions` format.
```
This last sentence is a bit confusing now.  The `expressions` router is always used, but the mode in which the configuration is specified can be changed between `expressions` and `traditional_compat`.  This PR only describes how the `expressions` router operates in `traditional_compat` mode.  Underneath, the routing specifications in are converted to the native `expressions` format.

Maybe we need to move this into a completely new section that describes the two routers ('traditional' and 'expressions'), the three modes ('traditional', 'traditional_compat', and 'expressions') and how they correlate.  With that out of the way, this description of `traditional_compat` mode will probably be less confusing.

@cloudjumpercat @kikito thoughts?

cloudjumpercat replied:
@hanshuebner I think I originally edited it to remove the word "underneath", so maybe that's why it's confusing now? If we wanted to add a sentence like "Underneath, the routing specifications in are converted to the native expressions format." back in, I'd try to use something like "At the ___ level," or something like that instead of "underneath".

> Maybe we need to move this into a completely new section that describes the two routers ('traditional' and 'expressions'), the three modes ('traditional', 'traditional_compat', and 'expressions') and how they correlate. With that out of the way, this description of traditional_compat mode will probably be less confusing.

I think if we wanted to do that, I almost think this content could be moved to a different page. I think the original intent of this "Routes Overview" page was just a general overview of routes and 'traditional', 'traditional_compat', and 'expressions' seem a lot more specific.

hanshuebner commented:
@cloudjumpercat I've reworded again to hopefully clarify.

cloudjumpercat reviewed (approved):
LGTM! Thanks for making these updates to this doc!
